pkg/component/component.go

```diff
@@ -367,104 +387,144 @@ func GetComponentSource(client *occlient.Client, componentName string, applicati
 		if err != nil {
-			return errors.Wrapf(err, "unable to update BuildConfig  for %s component", componentName)
-		}
-		err := Build(client, componentName, applicationName, false, true, stdout)
-		if err != nil {
-			return errors.Wrapf(err, "unable to build the component %v", componentName)
+			return errors.Wrapf(err, "unable to delete BuildConfig for %s component", componentName)
 		}
 
-		err = client.SetupForSupervisor(namespacedOpenShiftObject, projectName, annotations, labels)
+		// Update the DeploymentConfig
+		err = client.UpdateDCToSupervisor(commonObjectMeta, componentImageType)
 		if err != nil {
-			return errors.Wrapf(err, "unable to update DeploymentConfig  for %s component", componentName)
+			return errors.Wrapf(err, "unable to update DeploymentConfig for %s component", componentName)
 		}
+
 	} else {
 		// save source path as annotation
 		// this part is for updates where the source does not change or change from local to binary and vice versa
 
-		var annotations map[string]string
 		if newSourceType == "git" {
-			annotations = map[string]string{componentSourceURLAnnotation: newSource}
-			annotations[componentSourceTypeAnnotation] = newSourceType
+
+			// Update the BuildConfig
 			err = client.UpdateBuildConfig(namespacedOpenShiftObject, projectName, newSource, annotations)
 			if err != nil {
 				return errors.Wrapf(err, "unable to update the build config %v", componentName)
 			}
+
+			// Update DeploymentConfig annotations as well
 			err = client.UpdateDCAnnotations(namespacedOpenShiftObject, annotations)
 			if err != nil {
 				return errors.Wrapf(err, "unable to update the deployment config %v", componentName)
 			}
+
+			// Build it
 			err = Build(client, componentName, applicationName, true, true, stdout)
-		} else if newSourceType == "local" {
-			sourceURL := util.GenFileUrl(newSource, runtime.GOOS)
-			annotations = map[string]string{componentSourceURLAnnotation: sourceURL}
-			annotations[componentSourceTypeAnnotation] = newSourceType
-			err = client.UpdateBuildConfig(namespacedOpenShiftObject, projectName, "", annotations)
-			if err != nil {
-				return errors.Wrapf(err, "unable to update the build config %v", componentName)
-			}
-			err = client.UpdateDCAnnotations(namespacedOpenShiftObject, annotations)
-		} else if newSourceType == "binary" {
+
+		} else if newSourceType == "local" || newSourceType == "binary" {
+
+			// Update the sourceURL
 			sourceURL := util.GenFileUrl(newSource, runtime.GOOS)
-			annotations = map[string]string{componentSourceURLAnnotation: sourceURL}
-			annotations[componentSourceTypeAnnotation] = newSourceType
-			err = client.UpdateBuildConfig(namespacedOpenShiftObject, projectName, "", annotations)
-			if err != nil {
-				return errors.Wrapf(err, "unable to update the build config %v", componentName)
-			}
+			annotations[componentSourceURLAnnotation] = sourceURL
 			err = client.UpdateDCAnnotations(namespacedOpenShiftObject, annotations)
 		}
+
```

Comment:
can we move the line `err = client.UpdateDCAnnotations(namespacedOpenShiftObject, annotations)` here? instead of https://github.com/redhat-developer/odo/pull/622/files#diff-692effa1b0b7a9ed8c1f7bb49a977b9aR512 and https://github.com/redhat-developer/odo/pull/622/files#diff-692effa1b0b7a9ed8c1f7bb49a977b9aR525

Reply:
Can't because we use `build` before we update the annotations. Best to update the annotations before we start building or else it may cause errors if users use `odo create` or `odo update` during an update / build. If that makes sense.

Reply:
Hmm I see the build is triggered (only) if newSourceType is git